### ISSUE TITLE
'not' and 'between' condition bugs in lua 

### ISSUE THREAD

**Issue by Insolita** [yiisoft/yii2-redis#42]
``` php
        $q = JobRedis::find()->where(['between', 'id','5','10'])->limit(3)->asArray()->all();
       or 
        $q = JobRedis::find()->where(['between', 'id',5,10])->limit(3)->asArray()->all();
```

throw

```
[yii\db\Exception] Redis error: ERR Error running script (call to f_1c2ab0bf45b4b90789a37bff783c091c63700408): @user_script:10: user_script:10: attempt to compare string with boolean

Redis command was: EVAL local allpks=redis.call('LRANGE','job_redis',0,-1)
local pks={}
local n=0
local v=nil
local i=0
local key='job_redis'
for k,pk in ipairs(allpks) do
    local cid0=redis.call('HGET','job_redis' .. ':a:' .. pk, 'id')

    if cid0 >= '5' and cid0 <= '10' then
      i=i+1
      if i>0 and i<=3 then
        n=n+1 pks[n]=redis.call('HGETALL','job_redis:a:' .. pk)
      end
    end
end
return pks 0  
```

``` php
        $q = JobRedis::find()->where(['not', ['id'=>'5']])->limit(3)->asArray()->all();
        or         $q = JobRedis::find()->where(['not', ['id'=>5]])->limit(3)->asArray()->all();
        or         $q = JobRedis::find()->where(['not', ['id'=>null]])->limit(3)->asArray()->all();
```

throw

```
 [yii\db\Exception] Redis error: ERR Error compiling script (new function): user_script:10: unexpected symbol near '!' 
Redis command was: EVAL local allpks=redis.call('LRANGE','job_redis',0,-1)
local pks={}
local n=0
local v=nil
local i=0
local key='job_redis'
for k,pk in ipairs(allpks) do
    local cid0=redis.call('HGET','job_redis' .. ':a:' .. pk, 'id')

    if !(cid0=='5') then
      i=i+1
      if i>0 and i<=3 then
        n=n+1 pks[n]=redis.call('HGETALL','job_redis:a:' .. pk)
      end
    end
end
return pks 0  
```

**Comment by arogachev** [comment]
We're facing the same problem.

```php
$tasks = DeferredTask::find()->executable()->all();
```

```php
<?php

namespace common\modules\system\models\query;

use yii\redis\ActiveQuery;

class DeferredTaskQuery extends ActiveQuery
{
    /**
     * @return DeferredTaskQuery
     */
    public function executable()
    {   
        // <= is not supported, https://github.com/yiisoft/yii2-redis/issues/13    
        return $this->andWhere(['between', 'executed_at', '1970-01-01 00:00:00', date('Y-m-d H:i:s')]);
    }
}
```

After execution this exception is thrown:

```
'Redis error: ERR Error running script (call to f_ff6e83c21da2ee5dcda342b77d13bfd26fe082cf): @user_script:10: user_script:10: attempt to compare string with boolean
Redis command was: EVAL local allpks=redis.call('LRANGE','deferred_task',0,-1)
local pks={}
local n=0
local v=nil
local i=0
local key='deferred_task'
for k,pk in ipairs(allpks) do
    local cexecuted_at0=redis.call('HGET','deferred_task' .. ':a:' .. pk, 'executed_at')

    if cexecuted_at0 >= '1970-01-01 00:00:00' and cexecuted_at0 <= '2016-02-01 11:41:10' then
      i=i+1
      if i>0 then
        n=n+1 pks[n]=redis.call('HGETALL','deferred_task:a:' .. pk)
      end
    end
end
return pks 0'
```

I added a temporary workaround to made selection with PHP:

```php
$tasks = DeferredTask::find()->all();
foreach ($tasks as $task) {
    // Temporary replacement for scope executable()
    // https://github.com/yiisoft/yii2-redis/issues/42
    if (new DateTime($task->executed_at) > new DateTime()) {
        continue;
    }

    ...
}
```

**Important note:** But eariler `between` worked OK. We used Redis management GUI tool [Redis Desktop Manager](http://redisdesktop.com/) (Windows version) for manual deletion of some records, then we saw some records where all attributes have `null` values, maybe it's related to this problem.

**Comment by nailfor** [comment]
в файле LuaScriptBuilder нужно поменять возвращаемое условие:

    private function buildNotCondition($operator, $operands, &$params)
    {
        if (count($operands) != 1) {
            throw new InvalidParamException("Operator '$operator' requires exactly one operand.");
        }

        $operand = reset($operands);
        if (is_array($operand)) {
            $operand = $this->buildCondition($operand, $params);
        }

        //return "!($operand)";
        return "(not ($operand))";
    }

**Comment by cebe** [comment]
duplicate of #114 fixed by 4932c3cfd8dbd4fef88143430f1100ec767ba877
